File: airfield/service/instance.py
"""Service to manage zeppelin/jupyter instances"""

from datetime import datetime, timedelta
import random
import string
from ..storage.instance import InstanceStore
from ..adapter.marathon import MarathonAdapter, InstanceState
from .instance_zeppelin import ZeppelinInstanceService
from ..configuration.service import ConfigurationService
from ..settings import config
from ..util import dependency_injection as di
from ..util import metrics
from ..util.logging import logger
from ..api.auth import get_airfield_groups, get_user_groups
from ..util.exception import ConfigurationException, InstanceRunningTimeException


class InstanceService:
    @di.inject
    def __init__(self, zeppelin_instance: ZeppelinInstanceService, configuration_service: ConfigurationService,
                 instance_store: InstanceStore, marathon_adapter: MarathonAdapter):
        self._zeppelin_instance_service = zeppelin_instance
        self._configuration_service = configuration_service
        self._instance_store = instance_store
        self._marathon_adapter = marathon_adapter

    @metrics.instrument
    def get_instances(self, deleted=False):
        instances = list()
        instance_ids = self._instance_store.get_instance_ids(deleted=deleted)
        for instance_id in instance_ids:
            status = self.get_instance(instance_id, deleted=deleted)
            if status != {}:
                instances.append(status)
        return dict(instances=instances)

    @metrics.instrument
    def get_instance(self, instance_id, deleted=False):
        instance_configuration = self._instance_store.get_instance(instance_id, deleted=deleted)
        if not instance_configuration:
            return {}
        instance = self.get_instance_state(instance_id, instance_configuration=instance_configuration, deleted=deleted)
        instance["details"] = self._calculate_instance_details(instance_configuration)
        instance["proxy_url"] = "/proxy/{}".format(instance_id)
        return instance

    @metrics.instrument
    def get_instance_state(self, instance_id, instance_configuration=None, deleted=False):
        if instance_configuration is None:
            instance_configuration = self._instance_store.get_instance(instance_id, deleted=deleted)
        status = InstanceState.DELETED if deleted else self._marathon_adapter.get_instance_status(
            _instance_path(instance_configuration["configuration"], instance_id))
        stuck_deploying = False
        stuck_duration_seconds = 0
        if status == InstanceState.DEPLOYING:
            runtimes = instance_configuration["runtimes"]
            if runtimes and runtimes[-1]["stopped_at"] is None:
                started_at = datetime.fromtimestamp(runtimes[-1]["started_at"])
                if started_at + timedelta(seconds=10 * 60) <= datetime.now():
                    stuck_deploying = True
                    stuck_duration_seconds = (datetime.now() - started_at).seconds
        return {
            "instance_id": instance_id,
            "status": status.name,
            "deployment_stuck": stuck_deploying,
            "stuck_duration": stuck_duration_seconds
        }

    @metrics.instrument
    def get_instance_configuration(self, instance_id):
        instance_configuration = self._instance_store.get_instance(instance_id)
        return instance_configuration["configuration"]

    @metrics.instrument
    def get_instance_admins(self, instance_id):
        data = self._instance_store.get_instance(instance_id)
        return data["configuration"]["admin"]["admins"] + [data["metadata"]["created_by"]]

    @metrics.instrument
    def get_instance_url(self, instance_id):
        instance_configuration = self._instance_store.get_instance(instance_id)["configuration"]
        instance_path = _instance_path(instance_configuration, instance_id)
        ip, port = self._marathon_adapter.get_instance_ip_address_and_port(instance_path)
        if ip is not None and port is not None:
            return "{}:{}".format(ip, port)
        else:
            logger.error('Error while retrieving the ip address and the port of the instance {}'.format(instance_id))
            return None

    @metrics.instrument
    def get_instance_credentials(self, instance_id):
        instance_configuration = self._instance_store.get_instance(instance_id)["configuration"]
        if not instance_configuration["usermanagement"]["enabled"]:
            return []
        if instance_configuration["type"] == "zeppelin":
            return instance_configuration["usermanagement"]["users"]
        elif instance_configuration["type"] == "jupyter":
            return instance_configuration["usermanagement"]["password"]
        else:
            raise Exception("Unknown instance type: {}".format(instance_configuration["type"]))

    @metrics.instrument
    def get_instance_configurations(self):
        instances = list()
        instance_ids = self._instance_store.get_instance_ids()
        for instance_id in instance_ids:
            instance_configuration = self._instance_store.get_instance(instance_id)
            instances.append(dict(instance_id=instance_id, configuration=instance_configuration))
        return instances

    @metrics.instrument
    def create_instance(self, configuration, username):
        configuration = self._configuration_service.prepare_configuration(configuration)
        if config.OIDC_ACTIVATED and config.DCOS_GROUPS_ENABLED:
            group = configuration['admin']['group']
            if not group:
                raise ConfigurationException("No group is selected! Please select a group to deploy the instance!")
            if group not in get_user_groups():
                raise ConfigurationException(f"The user \"{username}\" is not authorized to deploy an instance in the group \"{group}\"!")
            if group not in get_airfield_groups():
                raise ConfigurationException(f"The group \"{group}\" is not available in airfield!")
        instance_id = _generate_instance_id()
        instance_path = _instance_path(configuration, instance_id)
        metadata = dict(created_by=username, created_at=datetime.now().timestamp())
        self._zeppelin_instance_service.create_instance(instance_path, configuration)
        configuration = self._instance_store.insert_instance(instance_id, configuration, metadata)
        self._start_runtime(instance_id, self._calculate_cost_factors(configuration))
        return instance_id

    @metrics.instrument
    def update_instance(self, instance_id, configuration):
        # The DCOS group of the instance is unchangeable, so we fetch the group from the already stored configuration.
        configuration["admin"]["group"] = self.get_instance_configuration(instance_id)["admin"]["group"]
        instance_path = _instance_path(configuration, instance_id)
        configuration = self._configuration_service.prepare_configuration(configuration)
        if self._marathon_adapter.get_instance_status(instance_path) != InstanceState.STOPPED:
            self._finish_runtime(instance_id)
        self._zeppelin_instance_service.update_instance(instance_path, configuration)
        configuration = self._instance_store.update_instance_configuration(instance_id, configuration)
        self._start_runtime(instance_id, self._calculate_cost_factors(configuration))
        return instance_id

    @metrics.instrument
    def delete_instance(self, instance_id):
        instance_configuration = self._instance_store.get_instance(instance_id)["configuration"]
        self._zeppelin_instance_service.delete_instance(_instance_path(instance_configuration, instance_id))
        # We do not check the status here, so the last runtime is always finished.
        self._finish_runtime(instance_id)
        self._add_deleted_at(instance_id)
        self._instance_store.delete_instance(instance_id)

    @metrics.instrument
    def start_instance(self, instance_id):
        instance_configuration = self._instance_store.get_instance(instance_id)
        instance_path = _instance_path(instance_configuration["configuration"], instance_id)
        if self._marathon_adapter.get_instance_status(instance_path) == InstanceState.STOPPED:
            self._marathon_adapter.start_instance(instance_path)
            self._start_runtime(instance_id, self._calculate_cost_factors(instance_configuration))
            return "started"
        return ""

    @metrics.instrument
    def stop_instance(self, instance_id):
        instance_configuration = self._instance_store.get_instance(instance_id)["configuration"]
        instance_path = _instance_path(instance_configuration, instance_id)
        if self._marathon_adapter.get_instance_status(instance_path) != InstanceState.STOPPED:
            self._marathon_adapter.stop_instance(instance_path)
            self._finish_runtime(instance_id)
            return "stopped"
        return ""

    @metrics.instrument
    def restart_instance(self, instance_id):
        instance_configuration = self._instance_store.get_instance(instance_id)
        instance_path = _instance_path(instance_configuration["configuration"], instance_id)
        if self._marathon_adapter.get_instance_status(instance_path) == InstanceState.STOPPED:
            self.start_instance(instance_id)
            return "started"
        self._marathon_adapter.restart_instance(instance_path)
        self._finish_runtime(instance_id)
        self._start_runtime(instance_id, self._calculate_cost_factors(instance_configuration))
        return "restarted"

    @metrics.instrument
    def get_default_configurations(self):
        return dict(configurations=self._configuration_service.get_available_configurations())

    def _calculate_cost_factors(self, instance_configuration):
        configuration = instance_configuration["configuration"]
        num_executors = int(configuration["spark"]["cores_max"]) / int(configuration["spark"]["executor_cores"])
        cores = int(configuration["notebook"]["cores"]) + num_executors * int(configuration["spark"]["executor_cores"])
        memory = int(configuration["notebook"]["memory"]) + num_executors * int(
            configuration["spark"]["executor_memory"])
        return dict(cores=cores, memory=memory)

    def _calculate_instance_details(self, instance_configuration):
        configuration = instance_configuration["configuration"]
        metadata = instance_configuration["metadata"]
        runtimes = instance_configuration["runtimes"]
        if runtimes and runtimes[-1]["stopped_at"] is None:
            started_at = datetime.fromtimestamp(runtimes[-1]["started_at"]).isoformat()
        else:
            started_at = None
        details = dict()
        details["comment"] = configuration["comment"]
        details["created_by"] = metadata["created_by"]
        for key in ["created_at", "deleted_at"]:
            if key in metadata:  # always true for "created_at"
                details[key] = datetime.fromtimestamp(metadata[key]).isoformat()
        details["running_since"] = started_at
        if config.COST_TRACKING_ENABLED:
            details["costs"] = _calculate_costs(instance_configuration)
        if configuration["delete_at"]:
            details["delete_at"] = datetime.fromtimestamp(configuration["delete_at"]).isoformat()
        return details

    def calculate_costs_per_hour(self, configuration):
        now = datetime.now().timestamp()
        instance_configuration = dict(configuration=configuration)
        cost_factors = self._calculate_cost_factors(dict(instance_configuration))
        cost_factors['started_at'] = now
        cost_factors['stopped_at'] = now + 3600

        instance_configuration['runtimes'] = [cost_factors]
        return dict(costs_per_hour=_calculate_costs(instance_configuration)['cost'])

    def _add_deleted_at(self, instance_id):
        metadata = self._instance_store.get_instance(instance_id, deleted=False)["metadata"]
        metadata["deleted_at"] = datetime.now().timestamp()
        self._instance_store.update_instance_metadata(instance_id, metadata)

    def _finish_runtime(self, instance_id):
        try:
            self._instance_store.finish_runtime(instance_id)
        except InstanceRunningTimeException as e:
            logger.debug(e)

    def _start_runtime(self, instance_id, cost_factors):
        try:
            self._instance_store.start_runtime(instance_id, cost_factors)
        except InstanceRunningTimeException as e:
            logger.debug(e)
            self._instance_store.finish_runtime(instance_id)
            self._instance_store.start_runtime(instance_id, cost_factors)


def _calculate_costs(instance_configuration):
    core_minutes = 0.0
    gb_minutes = 0.0
    running_time_seconds = 0
    for runtime in instance_configuration["runtimes"]:
        started_at = datetime.fromtimestamp(runtime["started_at"])
        stopped_at = datetime.fromtimestamp(runtime["stopped_at"]) if runtime["stopped_at"] else datetime.now()
        minutes = (stopped_at - started_at).seconds / 60
        running_time_seconds += (stopped_at - started_at).seconds
        core_minutes += runtime["cores"] * minutes
        gb_minutes += runtime["memory"] / 1024 * minutes
    cost = config.COST_CORE_PER_MINUTE * core_minutes + config.COST_GB_PER_MINUTE * gb_minutes
    return dict(cost=cost, running_time_seconds=running_time_seconds)


def _generate_instance_id() -> str:
    return ''.join(random.choice(string.ascii_lowercase + string.digits) for _ in range(9))


def _instance_path(configuration, instance_id):
    if configuration["admin"]["group"]:
           return "{}/{}".format(config.DCOS_GROUPS_MAPPING.get(configuration["admin"]["group"]), instance_id)
    else:
        return "{}/{}".format(config.MARATHON_APP_GROUP, instance_id)

File: frontend/mock/core/instance.js
/* eslint-disable camelcase */
const TimeService = require('../business/timeService');
const DefaultConfigurations = require('../ressource/defaultConfigurations');
const config = require('../config');
const notebooks = require('./notebook');
const store = require('../storage/instance');
const copyService = require('../business/copyService');

const costCorePerMinute = config.costCorePerMinute;
const costGBPerMinute = config.costGBPerMinute;
const costTrackingEnabled = config.costTrackingEnabled;
const defaultConfiguration = DefaultConfigurations[1].configuration;


const instanceStates = {
    HEALTHY: 'HEALTHY',
    STOPPED: 'STOPPED',
    DEPLOYING: 'DEPLOYING',
    UNHEALTHY: 'UNHEALTHY',
    NOT_FOUND: 'NOT_FOUND',
    DELETED: 'DELETED'
};


function getRandomArrayElement(array) {
    return array[Math.floor(Math.random() * array.length)];
}

function getRandomState() {
    return getRandomArrayElement(Object.values(instanceStates));
}

function generateId() {
    return Math.random().toString(36).replace(/[^a-z]+/g, '');
}

function prepareConfiguration(configuration){
    if(configuration.delete_at){
        configuration.delete_at = TimeService.toSeconds(configuration.delete_at);
    }
    if(configuration.type === 'zeppelin'){
        const users = configuration.usermanagement.users;
        Object.keys(users).forEach(username => {
            if(!users[username]){
                users[username] = generateId();
            }
        });
    }
    else if(configuration.type === 'jupyter'){
        if(!configuration.usermanagement.password){
            configuration.usermanagement.password = generateId();
        }
    }
    return configuration;
}

function calculateCosts(instance){
    let coreMinutes = 0.0;
    let gbMinutes = 0.0;
    let runningTimeSeconds = 0;
    instance.runtimes.forEach(runtime => {
        const startedAt = runtime.started_at;
        const stoppedAt = runtime.stopped_at || TimeService.now();
        runningTimeSeconds += stoppedAt - startedAt;
        const minutes = (stoppedAt - startedAt) / 60.0;
        coreMinutes += runtime.cores * minutes;
        gbMinutes += (runtime.memory / 1024) * minutes;
    });
    const cost = costCorePerMinute * coreMinutes + costGBPerMinute * gbMinutes;
    return {
        cost: cost,
        running_time_seconds: runningTimeSeconds
    };
}

function createRuntimeObject(instanceConfiguration, startedAt = undefined, stoppedAt = undefined){
    const numExecutors = parseInt(instanceConfiguration.spark.cores_max / instanceConfiguration.spark.executor_cores, 10);
    return {
        started_at: startedAt,
        stopped_at: stoppedAt,
        cores: instanceConfiguration.notebook.cores + numExecutors * instanceConfiguration.spark.executor_cores,
        memory: instanceConfiguration.notebook.memory + numExecutors * instanceConfiguration.spark.executor_memory
    };
}

function calculateInstanceDetails(instance){
    const lastRunTime = instance.runtimes[instance.runtimes.length - 1];
    const startedAt = !lastRunTime.stopped_at ? TimeService.toIsoString(lastRunTime.started_at) : undefined;
    const details = {
        comment: instance.configuration.comment,
        created_by: instance.metadata.created_by,
        running_since: startedAt
    };
    if(instance.configuration.delete_at){
        details.delete_at = TimeService.toIsoString(instance.configuration.delete_at);
    }
    ['created_at', 'deleted_at'].forEach(key => {
        // noinspection TypeScriptValidateTypes
        if(instance.metadata[key]){
            // noinspection TypeScriptValidateTypes
            details[key] = TimeService.toIsoString(instance.metadata[key]);
        }
    });
    if(costTrackingEnabled){
        details.costs = calculateCosts(instance);
    }
    return details;
}

function prepareInstance(instanceConfiguration, createdBy, createdAt){
    const id = generateId();
    instanceConfiguration = prepareConfiguration(instanceConfiguration);
    const instance = {
        instance_id: id,
        status: instanceStates.DEPLOYING,
        deployment_stuck: false,
        stuck_duration: 0,
        proxy_url: '/proxy/' + id,
        configuration: instanceConfiguration,
        metadata: {
            created_by: createdBy,
            created_at: createdAt
        }
    };
    instance.runtimes = [createRuntimeObject(instanceConfiguration, createdAt)];
    return instance;
}

function generateUsermanagement(){
    return {
        enabled: Math.random() < 0.5,
        password: Math.random() < 0.5 ? '123456' : null,
        users: Math.random() < 0.5 ? { admin: 'admin', guest1: '123', guest2: '', guest3: '' } : {}
    };
}

function createGeneratedInstance() {
    const now = TimeService.now();
    const instanceConfiguration = copyService.copy(defaultConfiguration);
    instanceConfiguration.type = Math.random() < 0.5 ? 'jupyter' : 'zeppelin';
    instanceConfiguration.comment = getRandomArrayElement([
        'Lorem ipsum dolor sit amet',
        'consectetur adipiscing elit, sed do eiusmod tempor',
        'foobar',
        ''
    ]);
    instanceConfiguration.usermanagement = generateUsermanagement();
    const createdAt = now - 120; // -120 makes debugging of detecting stuck deployments easier
    const instance = prepareInstance(instanceConfiguration, 'Host', createdAt);
    instance.status = getRandomState();
    instance.status = [instanceStates.DELETED, instanceStates.NOT_FOUND].includes(instance.status) ? instanceStates.HEALTHY : instance.status;
    if (instance.status === instanceStates.DEPLOYING){
        instance.stuck_duration = now - createdAt;
        instance.deployment_stuck = true;
    }
    if(instance.status === instanceStates.STOPPED){
        instance.runtimes[0].stopped_at = createdAt;
    }
    return instance;
}




module.exports = {
    getDefaultConfigurations(){
        return DefaultConfigurations;
    },
    calculateCostsPerHour(configuration){
        const instance = {
            configuration: configuration,
            runtimes: [createRuntimeObject(configuration, 0, 3600)]
        };
        return calculateCosts(instance).cost;
    },
    
    getPrices(){
        return {
            cost_tracking_enabled: costTrackingEnabled,
            cost_currency: 'EURO',
            cost_core_per_minute: costCorePerMinute,
            cost_gb_per_minute: costGBPerMinute
        };
    },
    
    __startRuntime(instanceId){
        const instance = this.getInstance(instanceId, false);
        this.__finishRuntime(instanceId); // finishes runtime if necessary
        store.addRuntime(instanceId, createRuntimeObject(instance.configuration, TimeService.now()));
    },
    
    __finishRuntime(instanceId){
        const instance = this.getInstance(instanceId);
        const len = instance.runtimes.length;
        const runtime = instance.runtimes[len - 1];
        runtime.stopped_at = runtime.stopped_at !== undefined ? runtime.stopped_at : TimeService.now();
        store.updateLastRuntime(instanceId, runtime);
    },
    
    updateInstance(instanceId, configuration){
        store.updateInstanceConfiguration(instanceId, prepareConfiguration(configuration));
        store.updateStatus(instanceId, instanceStates.HEALTHY);
        this.__startRuntime(instanceId);
        this.__finishRuntime(instanceId);
    },
    
    addInstance(configuration, createdBy){
        const group = configuration.admin.group;
        if(config.oidcActivated && config.dcosGroupsActivated && !group){
            // A group must be set for an instance, if not 409 is thrown.
            return ['No group is selected! Please select a group to deploy the instance!', 409];
        }
        const instance = prepareInstance(configuration, createdBy, TimeService.now());
        instance.status = instanceStates.HEALTHY;
        store.addInstance(instance, false);
        return['', 200];
    },
    
    getInstances(deleted = false){
        const instances = [];
        store.getInstances(deleted).forEach(instance => instances.push(this.getInstanceDetails(instance.instance_id, deleted)));
        return instances;
    },
    
    getInstance(instanceId, deleted = false) {
        return store.getInstance(instanceId, deleted) || {};
    },
    
    getInstanceDetails(instanceId, deleted = false){
        const instance = this.getInstance(instanceId, deleted);
        const instanceDetails = this.getStatus(instanceId, deleted);
        instanceDetails.proxy_url = instance.proxy_url;
        instanceDetails.details = calculateInstanceDetails(instance);
        return instanceDetails;
    },

    getStatus(instanceId, deleted = false) {
        let instance = {};
        try{
            instance = this.getInstance(instanceId, deleted);
        }
        catch (e){
            console.error(e);
            instance.status = instanceStates.NOT_FOUND;
        }
        const deploying = instance.status === instanceStates.DEPLOYING;
        instance.deployment_stuck = deploying ? Math.random() >= 0.5 : false;
        instance.stuck_duration = TimeService.formatRunningTime(deploying ? Math.round(Math.random() * 10) : 0);
        return {
            instance_id: instanceId,
            status: instance.status,
            stuck_duration: instance.stuck_duration,
            deployment_stuck: instance.deployment_stuck
        };
    },

    action(action, instanceId) {
        switch (action) {
            case 'start':
            case 'restart':
                store.updateStatus(instanceId, instanceStates.HEALTHY);
                this.__finishRuntime(instanceId);
                this.__startRuntime(instanceId);
                break;

            case 'stop':
                store.updateStatus(instanceId, instanceStates.STOPPED);
                this.__finishRuntime(instanceId);
                break;

            case 'delete':
                this.__finishRuntime(instanceId);
                const instance = this.getInstance(instanceId);
                store.updateStatus(instanceId, instanceStates.DELETED);
                instance.metadata.deleted_at = TimeService.now();
                store.addInstance(instance, true);
                store.deleteInstance(instanceId, false);
                break;
            default:
                console.log('Action not accepted: ' + action); // eslint-disable-line
                break;
        }
    },
    
    getInstanceCredentials(instanceId){
        const instanceConfiguration = this.getInstance(instanceId).configuration;
        let instanceCredentials = [];
        if(!(instanceConfiguration.usermanagement.enabled)){
            return instanceCredentials;
        }
        if(instanceConfiguration.type === 'zeppelin'){
            instanceCredentials = instanceConfiguration.usermanagement.users;
        }
        else if(instanceConfiguration.type === 'jupyter'){
            instanceCredentials = instanceConfiguration.usermanagement.password;
        }
        else {
            throw Error('Unknown instance type: ' + instanceConfiguration.type);
        }
        return instanceCredentials;
    }
};


for (let i = 0, l = Math.round(Math.random() * 10 + 5); i < l; ++i) {
    const instance = createGeneratedInstance();
    for (let j = 0, k = Math.round(Math.random() * 10 + 3); j < k; ++j){
        notebooks.addNotebook({ instanceId: instance.instance_id });
    }
    store.addInstance(instance, false);
}

console.info(`${store.getInstances(false).length} instances created.`);

